Assets/MixedRealityToolkit/Services/BaseEventSystem.cs

```diff
@@ -13,56 +13,209 @@ namespace Microsoft.MixedReality.Toolkit
     /// </summary>
     public abstract class BaseEventSystem : BaseService, IMixedRealityEventSystem
     {
-        #region IMixedRealityEventSystem Implementation
-
         private static int eventExecutionDepth = 0;
-        private readonly WaitUntil doneExecutingEvents = new WaitUntil(() => eventExecutionDepth == 0);
+        private readonly Type eventSystemHandlerType = typeof(IEventSystemHandler);
+
+        private List<(Type, IEventSystemHandler)> handlersToAdd = new List<(Type, IEventSystemHandler)>();
+        private List<(Type, IEventSystemHandler)> handlersToRemove = new List<(Type, IEventSystemHandler)>();
+
+        #region IMixedRealityEventSystem Implementation
 
         /// <inheritdoc />
         public List<GameObject> EventListeners { get; } = new List<GameObject>();
 
+        /// <inheritdoc />
+        public Dictionary<Type, HashSet<IEventSystemHandler>> HandlerEventListeners { get; } = new Dictionary<Type, HashSet<IEventSystemHandler>>();
+
         /// <inheritdoc />
         public virtual void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
         {
             Debug.Assert(!eventData.used);
+
+            HashSet<IEventSystemHandler> handlers;
+            if (!HandlerEventListeners.TryGetValue(typeof(T), out handlers))
+            {
+                return;
+            }
+
             eventExecutionDepth++;
 
-            for (int i = EventListeners.Count - 1; i >= 0; i--)
+            foreach (IEventSystemHandler handler in handlers)
             {
-                ExecuteEvents.Execute(EventListeners[i], eventData, eventHandler);
+                Debug.Assert(handler is T);
+                eventHandler.Invoke((T)handler, eventData);
             }
 
             eventExecutionDepth--;
+
+            if(eventExecutionDepth == 0 && (handlersToRemove.Count > 0 || handlersToAdd.Count > 0))
+            {
+                foreach(var handler in handlersToRemove)
+                {
+                    RemoveHandlerFromMap(handler.Item1, handler.Item2);
+                }
+
+                foreach (var handler in handlersToAdd)
+                {
+                    AddHandlerToMap(handler.Item1, handler.Item2);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void RegisterHandler<T>(IEventSystemHandler handler) where T : IEventSystemHandler
```

Comment:
I think it would be simpler to register handlers only under the type passed in here instead of once for each interface under `IEventSystemHandler` and, when handling an event, traverse all registered types and invoke their handlers if they can be assigned to the specific type that we're handling. 

That sounds more expensive than the current solution but I'd like to get an idea for how much before choosing the more complicated way. Could we maybe profile that in a release build?

Reply:
I did a brief profile of it, it seems to be ~2 times slower. Discussed it in person, it would also cause complications in other places. I kept the initial solution with registering multiple entries (if necessary) per handler.